src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -36,23 +36,30 @@ ClusterTaskManager::ClusterTaskManager(
       get_node_info_(get_node_info),
       announce_infeasible_task_(announce_infeasible_task),
       local_task_manager_(std::move(local_task_manager)),
-      scheduler_resource_reporter_(
-          tasks_to_schedule_, infeasible_tasks_, *local_task_manager_),
-      internal_stats_(*this, *local_task_manager_),
-      get_time_ms_(get_time_ms) {}
+      get_time_ms_(get_time_ms) {
+  if (local_task_manager_) {
+    scheduler_resource_reporter_ = std::make_unique<SchedulerResourceReporter>(
+        tasks_to_schedule_, infeasible_tasks_, *local_task_manager_);
+    internal_stats_ = std::make_unique<SchedulerStats>(*this, *local_task_manager_);
+  }
+}
 
 void ClusterTaskManager::QueueAndScheduleTask(
     const RayTask &task,
     bool grant_or_reject,
     bool is_selected_based_on_locality,
     rpc::RequestWorkerLeaseReply *reply,
-    rpc::SendReplyCallback send_reply_callback) {
+    rpc::SendReplyCallback send_reply_callback,
+    std::function<void(const NodeID &node_id)> node_selected_callback) {
   RAY_LOG(DEBUG) << "Queuing and scheduling task "
                  << task.GetTaskSpecification().TaskId();
+  auto callback = node_selected_callback
```

Comment:
we should RAY_CHECK that  send_reply_callback and node_selected_callback can't both set.